mongo/mongo.go

```diff
@@ -492,53 +522,50 @@ func ensureServer(args EnsureServerParams, mongoKernelTweaks map[string]string)
 		}
 	}
 
-	svcConf := newConf(ConfigArgs{
-		DataDir:       args.DataDir,
-		DBDir:         dbDir,
-		MongoPath:     mongoPath,
-		Port:          args.StatePort,
-		OplogSizeMB:   oplogSizeMB,
-		WantNUMACtl:   args.SetNUMAControlPolicy,
-		Version:       mongodVersion,
-		Auth:          true,
-		IPv6:          network.SupportsIPv6(),
-		MemoryProfile: args.MemoryProfile,
-	})
-	svc, err := newService(ServiceName, svcConf)
-	if err != nil {
-		return zeroVersion, errors.Trace(err)
-	}
-	installed, err := svc.Installed()
+	mongoArgs := generateConfig(mongoPath, args.DataDir, args.StatePort, oplogSize, args.SetNUMAControlPolicy, mongodVersion, args.MemoryProfile)
+	logger.Debugf("creating mongo service configuration for mongo version: %d.%d.%s at %q",
+		mongoArgs.Version.Major, mongoArgs.Version.Minor, mongoArgs.Version.Patch, mongoArgs.MongoPath)
+
+	svc, err := mongoArgs.asService()
 	if err != nil {
 		return zeroVersion, errors.Trace(err)
 	}
-	if installed {
-		exists, err := svc.Exists()
+
+	// TODO(tsm): refactor out to service.Configure
+	if featureflag.Enabled(feature.MongoDbSnap) {
+		err = mongoArgs.writeConfig(configPath(args.DataDir))
 		if err != nil {
 			return zeroVersion, errors.Trace(err)
 		}
-		if exists {
-			logger.Debugf("mongo exists as expected")
-			running, err := svc.Running()
-			if err != nil {
-				return zeroVersion, errors.Trace(err)
-			}
 
-			if !running {
-				return mongodVersion, errors.Trace(svc.Start())
-			}
-			return mongodVersion, nil
+		err := snap.SetSnapConfig(ServiceName, "configpath", configPath(args.DataDir))
+		if err != nil {
+			return zeroVersion, errors.Trace(err)
 		}
+
+		err = service.ManuallyRestart(svc)
+		if err != nil {
+			logger.Criticalf("unable to (re)start mongod service: %v", err)
+			return zeroVersion, errors.Trace(err)
+		}
+
+		return mongodVersion, nil
 	}
 
-	if err := svc.Stop(); err != nil {
-		return zeroVersion, errors.Annotatef(err, "failed to stop mongo")
+	running, err := svc.Running()
+	if err != nil {
+		return zeroVersion, errors.Trace(err)
+	}
+	if running {
+		return mongodVersion, nil
 	}
+
+	dbDir := DbDir(args.DataDir)
 	if err := makeJournalDirs(dbDir); err != nil {
-		return zeroVersion, fmt.Errorf("error creating journal directories: %v", err)
+		return Version{}, errors.Errorf("error creating journal directories: %v", err)
 	}
-	if err := preallocOplog(dbDir, oplogSizeMB); err != nil {
-		return zeroVersion, fmt.Errorf("error creating oplog files: %v", err)
+	if err := preallocOplog(dbDir, args.OplogSize); err != nil {
```

Comment:
Ooops, we have set oplogSizeMB above and are not using it. The change here is a mistake it seems.
Also, why the sudden aversion to zeroVersion?

Reply:
Good catch

I'll get `zeroVersion` in there.